templates/base/client/.hathora/client.ts.hbs

```diff
@@ -122,62 +106,89 @@ export class HathoraConnection {
   private changedAt = 0;
   private updateListeners: UpdateCallback[] = [];
   private errorListeners: ErrorCallback[] = [];
+  private transport: HathoraTransport;
+  private internalState: UserState | undefined;
 
   constructor(
+    public appId: string,
     public stateId: StateId,
-    public state: UserState,
-    private socket: WebSocket,
+    transportType: TransportType,
     onUpdate?: UpdateCallback,
     onError?: ErrorCallback
   ) {
+    if (transportType === TransportType.TCP) {
+      this.transport = new TCPHathoraTransport(appId);
+    } else if (transportType === TransportType.WebSocket) {
+      this.transport = new WebSocketHathoraTransport(appId);
+    } else {
+      throw new Error("Unknown transport type");
+    }
+
     if (onUpdate !== undefined) {
       this.onUpdate(onUpdate);
-      onUpdate({ stateId, state: JSON.parse(JSON.stringify(this.state)), updatedAt: 0, events: [] });
     }
     if (onError !== undefined) {
       this.onError(onError);
     }
-    socket.onmessage = ({ data }) => {
-      const reader = new Reader(new Uint8Array(data as ArrayBuffer));
-      const type = reader.readUInt8();
-      if (type === 0) {
-        this.state = decodeStateSnapshot(reader);
-        this.changedAt = 0;
-        this.updateListeners.forEach((listener) =>
-          listener({ stateId, state: JSON.parse(JSON.stringify(this.state)), updatedAt: 0, events: [] })
-        );
-      } else if (type === 1) {
-        const { stateDiff, changedAtDiff, responses, events } = decodeStateUpdate(reader);
-        if (stateDiff !== undefined) {
-          this.state = computePatch(this.state!, stateDiff);
-        }
-        this.changedAt += changedAtDiff;
-        this.updateListeners.forEach((listener) =>
-          listener({
-            stateId,
-            state: JSON.parse(JSON.stringify(this.state)),
-            updatedAt: this.changedAt,
-            events: events.map((e) => e.event),
-          })
-        );
-        responses.forEach(({ msgId, response }) => {
-          if (msgId in this.callbacks) {
-            this.callbacks[msgId](response);
-            delete this.callbacks[msgId];
-          }
-        });
-      } else if (type === 2) {
-        this.disconnect(4004);
-      } else {
-        console.error("Unknown message type", type);
-      }
-    };
-    socket.onclose = (e) => {
-      console.error("Connection closed", e);
-      this.errorListeners.forEach((listener) => listener(transformCoordinatorFailure(e)));
-    };
   }
 
+  public async connect(token: string): Promise<void> {
+    this.internalState = await this.transport.connect(this.stateId, token, this.handleData, this.handleClose);
+    this.updateListeners.forEach((updateListener) =>
+      updateListener({ stateId: this.stateId, state: this.state, updatedAt: 0, events: [] })
```

Comment:
why special case the initial state? can't the transport just call `onData` instead of returning anything?